Docker and GKE container support.

qingling128 opened the pull request.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -38,12 +38,14 @@ class Map
-    # Constants for service names and resource types.
+    # Constants.
     module Constants
+      # Service names and resource types.
```
Makes sense. Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -11,9 +11,11 @@
 require 'grpc'
 require 'json'
 require 'open-uri'
+require 'rubygems'
```
Ah I see. Good to know.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -58,7 +60,11 @@ module Constants
+        metadata_attributes: %w(kube-env)
+      }
+      DOCKER_CONSTANTS = {
+        service: 'dockercontainer.googleapis.com',
```
There is no existing service with this name. Logging API V2 no longer replies `XXX.googleapis.com` type of services. So the purpose of this is just to provide a unique namespace for the labels. I've just changed it to `docker.googleapis.com`. BTW I had sent the backend folks an email regarding this naming a while ago. Will resume conversation on that thread as well.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -154,6 +160,26 @@ module Constants
+
+      # Use empty string as request path when locally-unique key of monitored
+      # resource can be implicitly inferred by Metadata Agent.
+      IMPLICIT_MONITORED_RESOURCE_UNIQUE_KEY = ''
```
Renamed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -154,6 +160,26 @@ module Constants
+      DEFAULT_UNIQUE_ID_KEY = "#{DEFAULT_PAYLOAD_KEY_PREFIX}/locally_unique_id"
+
+      # Docker container support.
+      DEFAULT_DOCKER_API_SOCKET_PATH = '/var/run/docker.sock'
```
Oops. Removed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -401,7 +440,7 @@ def configure(conf)
       # Log an informational message containing the Logs viewer URL
       @log.info 'Logs viewer address: https://console.cloud.google.com/logs/',
-                "viewer?project=#{@project_id}&resource=#{@resource_type}/",
+                "viewer?project=#{@project_id}&resource=#{@resource.type}/",
```
I changed it to:
```
      @log.info 'Logs viewer address: https://console.cloud.google.com/logs/',
                "viewer?project=#{@project_id}"
```

This directs users to the project main logging page. They will need to select the resource type and filters themselves. Is this what we want?

qingling128 replied:
I created b/65257345 for this follow up.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -417,245 +456,243 @@ def shutdown
+              determine_entry_level_monitored_resource_and_labels(
+                group_level_resource, group_level_common_labels, record)
+
+            if [CONTAINER_CONSTANTS[:resource_type],
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -417,245 +456,243 @@ def shutdown
+            if [CONTAINER_CONSTANTS[:resource_type],
+                DOCKER_CONSTANTS[:resource_type]].include?(
+                  entry_level_resource.type)
+              # Save the timestamp if available, then clear it out to allow for
```
Sounds good.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1112,14 +1218,57 @@ def determine_entry_level_labels(group_resource, record)
+      )
+    end
+
+    # Call Metadata Agent and parse response to json. Return nil in case of any
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1112,14 +1218,57 @@ def determine_entry_level_labels(group_resource, record)
+
+    # Call Metadata Agent and parse response to json. Return nil in case of any
+    # error / failure.
+    def call_metadata_agent(path)
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1112,14 +1218,57 @@ def determine_entry_level_labels(group_resource, record)
+      end
+    rescue StandardError => e
+      @log.error 'Error calling Metadata Agent.', error: e
+      return nil
```
Removed `return nil`.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1160,11 +1334,20 @@ def setup_prometheus
     Prometheus::Client.registry.instance_variable_set('@metrics', {})
   end
 
+  def setup_metadata_agent_stubs(resource_type)
```
As I understand the way webmock works, we can stub out all the possible URLs at once (in fact, we could even have a static unique_id->monitored_resource map, and use [dynamic response generation](https://github.com/bblimke/webmock#responses-dynamically-evaluated-from-lambda)). We can then set expectations for specific URLs being requested in individual tests.

qingling128 replied:
Refactored. PTAL.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -154,6 +160,26 @@ module Constants
+      # The label name of locally unique id in the json payload. When a record
+      # has this field in the payload, we will use the value to retrieve
+      # monitored resource from Stackdriver Metadata agent.
+      DEFAULT_UNIQUE_ID_KEY = "#{DEFAULT_PAYLOAD_KEY_PREFIX}/locally_unique_id"
```
This would probably problematic to change at this point, but I just realized that this string does not mention the word "resource" anywhere. It would make more sense to call this "#{DEFAULT_PAYLOAD_KEY_PREFIX}/local_resource_id" instead. WDYT?

qingling128 replied:
I think `local_resource_id` makes more sense. And it's not too late to change it at this point. So I just changed it along with all the comments. I'll change the configs in the EAP doc once I push a new gem.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -154,6 +160,26 @@ module Constants
+      # The label name of locally unique id in the json payload. When a record
+      # has this field in the payload, we will use the value to retrieve
+      # monitored resource from Stackdriver Metadata agent.
+      DEFAULT_UNIQUE_ID_KEY = "#{DEFAULT_PAYLOAD_KEY_PREFIX}/locally_unique_id"
```
Why do we call this `DEFAULT_UNIQUE_ID_KEY`? Can this key be overridden? If not, let's call it something like `LOCAL_RESOURCE_ID_KEY` or something...

qingling128 replied:
Nope, this can not be overridden. I just changed to `LOCAL_RESOURCE_ID_KEY`.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,180 @@ def test_log_entry_trace_field
+    assert_true d.instance.instance_variable_get(:@enable_metadata_agent)
+  end
+
+  # Test an implicit monitored resource can be retrieved from Metadata Agent
```
Test *that* an implicit ...

qingling128 replied:
Changed.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,180 @@ def test_log_entry_trace_field
+    end
+  end
+
+  # Test the enable_metadata_agent set to true.
```
Changed.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,180 @@ def test_log_entry_trace_field
 
+  # Metadata Agent related tests.
+
+  # Test the enable_metadata_agent config not set or set to false.
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1227,9 +1376,14 @@ def compute_severity(resource_type, record, entry_common_labels)
+        stream = entry_level_common_labels[
+          "#{CONTAINER_CONSTANTS[:service]}/stream"] if
+           resource_type == CONTAINER_CONSTANTS[:resource_type]
+        stream = entry_level_common_labels[
```
Right. I've removed this logic from the code base. For now the configuration for docker container stderr/stdout logs looks like this:
```
# For Docker container stderr / stdout logs.
<filter container.**>
  @type record_transformer
  enable_ruby
  <record>
    "logging.googleapis.com/locally_unique_id" ${tag}
    "severity" ${record['source']}
  </record>
  remove_keys container_id,container_name
</filter>
```

While the command to run docker looks like:
```
sudo docker run -it --log-driver=fluentd --log-opt fluentd-address=tcp://127.0.0.1/24224 --log-opt tag="container.{{.FullID}}"
```

Note that to support the additional `severity` field for docker container, we will need to save then restore `severity` field just like what we did to `time` when parsing json. (This was not an issue for GKE.)

igorpeshansky replied:
As I said elsewhere, can't we just set the severity explicitly, e.g.:
```
    "severity" ${if record['source']=='STDOUT' then 'INFO' else 'ERROR' end}
```
?

qingling128 replied:
Right. I've removed the logic to map `severity` for `stderr` and `stdout`.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1100,6 +1197,15 @@ def determine_entry_level_labels(group_resource, record)
+        # 'source' from the json record and set it as a common label.
+        common_labels.merge!(
+          delete_and_extract_labels(
+            record, 'source' => "#{DOCKER_CONSTANTS[:service]}/stream"))
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,81 +993,90 @@ def determine_agent_level_common_labels
+      # NO_UNIQUE_ID_PRESENT.
+      groups[NO_UNIQUE_ID_PRESENT] = []
+      entries.each do |time, record|
+        if record.is_a?(Hash) && record.key?(DEFAULT_UNIQUE_ID_KEY)
```
It *should* always be a `Hash`. User could mess with it though. For example they can send non-hash values via `in_forward` plugin. We used to have this check. So I kept it. We should add some logs when it's not a hash indicating we are dropping invalid log entries though.

qingling128 replied:
There is a test `test_malformed_log` that makes sure we throw away non-hash logs instead of letting the agent crash. For logs coming from `in_forward` plugin, the `record` can be anything.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,81 +993,90 @@ def determine_agent_level_common_labels
+      groups = {}
+      # Log entries with no locally unique id go into a group with key
+      # NO_UNIQUE_ID_PRESENT.
+      groups[NO_UNIQUE_ID_PRESENT] = []
```
This no longer exists after commit e0cdf75caeece1a9e80ae4274280d6d73f629cda. Instead we are hashing by the tag & local_resource_id pairs.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -417,245 +456,243 @@ def shutdown
-            end
-            if record.length == 1 && record.key?('message')
-              record_json = parse_json_or_nil(record['message'])
+      entries_grouped_by_tag = group_log_entries_by_tag(chunk)
```
This has been changed in commit e0cdf75caeece1a9e80ae4274280d6d73f629cda.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -154,6 +160,26 @@ module Constants
       }
+
+      # Used as a map key for entries with no locally unique id in the payload.
+      NO_UNIQUE_ID_PRESENT = 'no_locally_unique_id'
```
Yup, this has been removed as in commit e0cdf75caeece1a9e80ae4274280d6d73f629cda.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -351,22 +382,30 @@ def configure(conf)
+        # The locally-unique key for this should be the instance id. Since this
+        # can be implicitly inferred by Metadata Agent, we do not need to
+        # explicitly send the key.
+        @resource = call_metadata_agent_for_monitored_resource(
```
We've been doing this for a while. (See details in `determine_agent_level_monitored_resource_via_legacy`). I've created a separate ticket for potential refactoring: b/65175256.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1485,7 +1638,8 @@ def set_payload(resource_type, record, entry, is_json)
         entry.json_payload = record
-      elsif resource_type == CONTAINER_CONSTANTS[:resource_type] &&
+      elsif [CONTAINER_CONSTANTS[:resource_type],
+             DOCKER_CONSTANTS[:resource_type]].include?(resource_type) &&
```
Yeah, that sounds good to me.

qingling128 commented:
All feedback addressed. PTAL. (I noticed there are some conflicts with master. I'll hold for rebasing at this moment. Will wait for https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/146 to be merged first).

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,188 @@ def test_log_entry_trace_field
+    end
+  end
+
+  # Test logs from applications running in GKE. These logs has the label
```
Nit: `These logs have the label`... But I wonder why you chose to call it out for GKE, but not for Docker...

qingling128 replied:
Changed `has` to `have` and called it out for Docker container as well.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1160,11 +1342,27 @@ def setup_prometheus
     Prometheus::Client.registry.instance_variable_set('@metrics', {})
   end
 
+  def url_with_local_resource_id(local_resource_id)
```
`metadata_request_url`?

qingling128 replied:
Changed.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1160,11 +1342,27 @@ def setup_prometheus
       "#{container_name}"
   end
 
+  # GKE Container
```
(a) Should this be above `container_tag_with_container_name`?
(b) Missing trailing '.'.

qingling128 replied:
Moved this above and added `.`.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1160,11 +1342,27 @@ def setup_prometheus
     Prometheus::Client.registry.instance_variable_set('@metrics', {})
   end
 
+  def url_with_local_resource_id(local_resource_id)
```
Do we want to put a `# Metadata Agent.` comment above these functions?

qingling128 replied:
Added.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1192,6 +1390,34 @@ def container_log_entry(log, stream = CONTAINER_STREAM)
+    }
+  end
+
+  # Docker Container
```
Nit: Missing trailing '.'.

qingling128 replied:
Added.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1192,6 +1390,34 @@ def container_log_entry(log, stream = CONTAINER_STREAM)
+  end
+
+  # Docker Container
+  def docker_container_stdout_stderr_log_entry(
```
[Optional] `docker_container_stream_log_entry`...

qingling128 replied:
Hmm... I think `stderr` and `stdout` is more straight forward.

igorpeshansky replied:
Fine, but can we at least be consistent about the ordering of `stdout_stderr` vs `stderr_stdout`?

qingling128 replied:
Makes sense. I changed all occurrences to `stdout_stderr`.

igorpeshansky commented on test/plugin/constants.rb:
```diff
@@ -23,8 +24,10 @@ module Constants
   # Attributes used for the GCE metadata service.
   PROJECT_ID = 'test-project-id'
   ZONE = 'us-central1-b'
+  ZONE2 = 'us-central1-c'
```
Let's call these `METADATA_ZONE` and `METADATA_VM_ID`, to make it consistent with `CUSTOM_ZONE` and `CUSTOM_VM_ID` below...

qingling128 replied:
Changed.

igorpeshansky commented on test/plugin/constants.rb:
```diff
@@ -477,4 +562,50 @@ module Constants
+  MONITORED_RESOURCE_STUBS = {
+    # Implicit GCE instance.
+    IMPLICIT_LOCAL_RESOURCE_ID =>
+      '{' \
```
How about:
```ruby
  MONITORED_RESOURCE_STUBS = {
    # Implicit GCE instance.
    IMPLICIT_LOCAL_RESOURCE_ID =>
      {
        'type' => COMPUTE_CONSTANTS[:resource_type],
        'labels' => {
          'zone': ZONE2,
          'instance_id': VM_ID2
        }
      }.to_json,
    # Docker container stderr / stdout logs.
    "container.#{DOCKER_CONTAINER_ID}" =>
      {
        ...
      }.to_json,
      ...
  }
```
?

qingling128 replied:
Good idea. Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -105,6 +112,23 @@ module Constants
+
+      # Use empty string as request path when locally-unique id of monitored
+      # resource can be implicitly inferred by Metadata Agent.
+      IMPLICIT_MONITORED_RESOURCE_UNIQUE_ID = ''
```
Moved.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -105,6 +112,23 @@ module Constants
+    end
+
+    # Constants for log entry field extraction.
+    module LogEntryFieldsConstants
```
Renamed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -463,16 +490,21 @@ def write(chunk)
             unless record.key?('time') || timestamp.nil?
               record['time'] = timestamp
             end
+            unless record.key?('severity') || severity.nil?
```
Good idea. Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
+          next
+        end
+        locally_unique_id = record.delete(DEFAULT_UNIQUE_ID_KEY) if
+          record.is_a?(Hash)
```
For logs from `in_forward` plugin, the record can be anything. We even have a test `test_malformed_log` to test this case. I added some logging for it.

qingling128 replied:
Cool cool.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
+        end
+        locally_unique_id = record.delete(DEFAULT_UNIQUE_ID_KEY) if
+          record.is_a?(Hash)
+        hash_key = [sanitized_tag, locally_unique_id].freeze
```
Ah, I removed it in order to leave it as `nil` so that later on we can evaluate `if local_resource_id` as false. If I convert it to an empty string, it becomes confusing as we also have the implicit local resource id as an empty string.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
+      resource = @resource.dup
+      resource.labels = @resource.labels.dup
+      common_labels = @common_labels.dup
+      matched_regexp_group = nil
```
Yes. if it's not `GKE`, the `@tag_regexp_list` will be empty. I added this line because some tests were failing.

igorpeshansky replied:
How would you feel about calling it out explicitly with a "hideously awesome" recipe I [found somewhere](https://stackoverflow.com/a/13299030/6752892):
```ruby
      if @tag_regexp_list.each do |derived_type, tag_regexp|
        matched_regexp_group = tag_regexp.match(tag)
        if matched_regexp_group
          resource.type = derived_type
          break
        end
      end.empty?
        matched_regexp_group = nil
      end
```
? Or with something less aw{esome,ful} and more conventional, e.g.:
```ruby
      unless @tag_regexp_list.empty?
        @tag_regexp_list.each do |derived_type, tag_regexp|
          matched_regexp_group = tag_regexp.match(tag)
          if matched_regexp_group
            resource.type = derived_type
            break
          end
        end
      else
        matched_regexp_group = nil
      end
```
?

qingling128 replied:
The first one kind of scared me...

We would need to initiate `matched_regexp_group` in the beginning because of how ruby decides its variable scope. As shown below `if` works, but `for each loop` does not.

```
$ irb
2.1.10 :002 > if true
2.1.10 :003?>     foo = 'a'
2.1.10 :004?>   else
2.1.10 :005 >       foo = 'b'
2.1.10 :006?>   end
 => "a"
2.1.10 :007 > foo
 => "a"
2.1.10 :008 > {'a' => 'a'}.each do |k, v|
2.1.10 :009 >       val = 'val'
2.1.10 :010?>   end
 => {"a"=>"a"}
2.1.10 :011 > val
NameError: undefined local variable or method `val' for main:Object
        from (irb):11
        from /usr/local/google/home/lingshi/.rvm/rubies/ruby-2.1.10/bin/irb:11:in `<main>'
```

I changed the code to be
```
      matched_regexp_group = nil
      # @tag_regexp_list can be an empty list.
      @tag_regexp_list.each do |derived_type, tag_regexp|
        matched_regexp_group = tag_regexp.match(tag)
        if matched_regexp_group
          resource.type = derived_type
          break
        end
      end
```
if that helps.

igorpeshansky replied:
Ah, I didn't think about the block scoping issue. What you have now works, unless you also want to move the comment to the end of the previous line, i.e.:
```ruby
      matched_regexp_group = nil  # @tag_regexp_list can be an empty list.
```

qingling128 replied:
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
-      group_resource_labels = @resource.labels.dup
-      group_common_labels = @common_labels.dup
+      # Determine the monitored resource based on the locally unique id.
+      # Different monitored resource type has unique ids in different format. We
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
+      # Different monitored resource type has unique ids in different format. We
+      # will query Metadata Agent for the monitored resource. Return the
+      # original monitored resource (either the instance resource or the
+      # resource inferred by tag) if failing to get a monitored resource from
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
+        @log.debug 'Retrieved monitored resource from metadata agent: ' \
+                  "#{retrieved_resource.inspect}."
+        unless retrieved_resource.nil?
+          # Temporarily renaming 'gke_container' to 'container'.
```
Added.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +986,166 @@ def determine_agent_level_common_labels
         end
+
+      # Docker container.
+      when DOCKER_CONSTANTS[:resource_type]
```
Per discussion with Summit, we should not be sending this metadata label for docker container since it's been ingested in the Logging Backend. For other existing resource types we have been doing it for a while and could not easily deprecate it.

igorpeshansky replied:
> *Aside*: I feel we're introducing more legacy stuff to get rid of by propagating `DOCKER_CONSTANTS` throughout the code. Ideally, we shouldn't need to reference it at all -- the code should not care what the resource looks like.

That said, in this particular case we shouldn't need to remove the `resource_name` field because it would never be added (lines 975-976 should only be executed on raw GCE or on GKE). So this seems like dead code to me, unless I'm missing something.

qingling128 replied:
The line 975-976 gets executed on agent startup. At that point we can't tell `a raw GCE instance` from `a GCE instance with docker installed and potentially want to send docker logs`. The we will be adding this `resource_name` label regardless.

`Docker container` logs only get identified when the monitored resource (specified by `logging.googleapis.com/local_resource_id`) returned by metadata agent has `resource_type` as `docker_container`. This process happens kind of late when logs start coming in.

In a word, we still need to remove `resource_name` for `docker_container` unless we change the behavior dramatically as proposed in b/65175256.

igorpeshansky replied:
Fine, makes sense. Can we add a TODO to remove this once the resource is retrieved at a proper time?

qingling128 replied:
Added.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1324,6 +1430,9 @@ def set_labels(record, entry)
       'C' => 'CRITICAL',
       'A' => 'ALERT',
+      # Container stderr / stdout logs.
+      'STDOUT' => 'INFO',
```
I've removed this map. We will address the `severity` value in the config.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,188 @@ def test_log_entry_trace_field
+      verify_log_entries(n, DOCKER_CONTAINER_PARAMS)
+      assert_requested_metadata_agent_stub(IMPLICIT_LOCAL_RESOURCE_ID)
+      assert_requested_metadata_agent_stub("container.#{DOCKER_CONTAINER_ID}")
+      WebMock.reset!
```
I changed `setup_metadata_agent_stubs` to take a block. The naming or the method is left as it is, similar to `setup_logging_stubs`.

igorpeshansky replied:
Careful, `WebMock` covers all stubs, not just the metadata agent stubs. In fact, ideally, this would be part of the `setup_logging_stubs` context, because it also changes `@logs_sent`, but, unfortunately, we want to stop mocking the logging API early, to ensure `d.emit` is only called within that block... I wish Ruby allowed passing multiple blocks (one to execute the code and one to verify the results), but, alas, this seems to be [woefully inefficient](https://mudge.name/2011/01/26/passing-blocks-in-ruby-without-block.html)...
So, my suggestion was to create a top-level context object to wrap all stubs. Does that sound reasonable?

qingling128 replied:
Yes, `WebMock.reset!` will reset all stubs. I had to switch the order of a few lines to make sure that did not break out tests. If we are doing a top-level context object, we will be setting up **all** stubs for **all** tests that need any stub, right? That would make our test a bit cleaner. I wonder if it is a bit overkill though.

On the other hand, if we decide to do that, we should probably add `assert_expected_stub` checks for each case when a stub is used.

WDYT?

igorpeshansky replied:
Oh, sorry I wasn't clear. I was simply suggesting a stub context object that would initialize `@logs_sent` and reset `WebMock` at the end. We could still call `setup_*_stubs` in the block passed into the context to set up only the stubs we need. The `assert_expected_stub` calls you have now would also have to appear within the block (before the reset). See my earlier example... Does this make more sense?

qingling128 replied:
AH! I see what you meant now. Sorry I misunderstood. Changed in a82a36334b933299a81a058eaf91cddd32d0e2e7.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,188 @@ def test_log_entry_trace_field
+      # Timestamp in 'time' field from log entry should be set properly.
+      assert_equal DOCKER_CONTAINER_SECONDS_EPOCH,
+                   entry['timestamp']['seconds'], entry
+      assert_equal DOCKER_CONTAINER_NANOS, entry['timestamp']['nanos'],
```
Done.

qingling128 commented on test/plugin/base_test.rb:
```diff
@@ -1045,6 +1045,188 @@ def test_log_entry_trace_field
+  # Test that the 'source' field is properly extracted from the record json and
+  # set as a common label 'stream'. Also entry.severity is set accordingly for
+  # Docker container logs.
+  def test_source_and_severity_for_docker_container_logs
```
Right. I've removed the `severity` check.

igorpeshansky replied:
Should we not rename the test as well?

qingling128 replied:
I guess you meant `Should we rename the test as well?` Renamed it to `test_source_for_docker_container_logs` and also fixed the comments.

igorpeshansky replied:
I meant "shouldn't we". :-) But thanks.

qingling128 replied:
You're welcome.

qingling128 commented:
PTAL.

The requested changes that requires a `git rebase` from `master` will be postponed till everything else seems fine. Because I don't want to confuse GitHub and mess up the comment history (learnt that the hard way).

qingling128 commented:
I replied to the last two comments. But for some reason it does not show up in the `conversation` tab... So let me answer them here.

[The `record.is_a?(Hash)` check]
There is a test `test_malformed_log` that makes sure we throw away `non-hash` logs instead of letting the agent crash. For logs coming from `in_forward` plugin, the record can be anything. Previously we are checking it in two places. I just modified it to check it once.

[`severity` mapping logic]
I've removed the logic to map `severity` for `stderr` and `stdout`. We will rely on configurations instead.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +983,172 @@ def determine_agent_level_common_labels
+          next
+        end
+        local_resource_id = record.delete(LOCAL_RESOURCE_ID_KEY)
+        # If local_resource_id is nil, use that as the hash key anyway.
```
I'd put it another way: `# A nil local_resource_id means "fall back to legacy".`

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +983,172 @@ def determine_agent_level_common_labels
+      chunk.msgpack_each do |tag, time, record|
+        unless record.is_a?(Hash)
+          @log.warn "Dropping log entries with malformed record: '#{record}'." \
+                    ' A log record should be in json format.'
```
Nit: JSON.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +983,172 @@ def determine_agent_level_common_labels
+      groups = {}
+      chunk.msgpack_each do |tag, time, record|
+        unless record.is_a?(Hash)
+          @log.warn "Dropping log entries with malformed record: '#{record}'." \
```
[Optional] Do we want `#{record.inspect}`? Or is a stringified version of the record sufficient to figure out what it was?

qingling128 replied:
Guess it does not hurt using `#{record.inspect}` (only better). Just changed it.

igorpeshansky commented on test/plugin/constants.rb:
```diff
@@ -215,7 +244,16 @@ module Constants
       "#{COMPUTE_CONSTANTS[:service]}/resource_name" => HOSTNAME
     }
   }
+  COMPUTE_PARAMS_WITH_DIFFERENT_VM_ID_AND_ZONE = COMPUTE_PARAMS.merge(
```
`COMPUTE_PARAMS_WITH_METADATA_VM_ID_AND_ZONE`

qingling128 replied:
Changed.

igorpeshansky commented on test/plugin/constants.rb:
```diff
@@ -477,4 +563,50 @@ module Constants
+          'instance_id' => METADATA_VM_ID
+        }
+      }.to_json,
+    # Docker container stderr g stdout logs.
```
I think you meant to keep the `/` here instead of `g`.

qingling128 replied:
LOL. Oops. Changed.

igorpeshansky commented on test/plugin/constants.rb:
```diff
@@ -23,8 +25,10 @@ module Constants
   # Attributes used for the GCE metadata service.
   PROJECT_ID = 'test-project-id'
   ZONE = 'us-central1-b'
+  METADATA_ZONE = 'us-central1-c'
```
I would pull these out into their own section, e.g.:
```ruby
  # Attributes used for the Metadata Agent resources.
  METADATA_ZONE = 'us-central1-c'
  METADATA_VM_ID = '0123456789'
```

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +983,174 @@ def determine_agent_level_common_labels
+      chunk.msgpack_each do |tag, time, record|
+        unless record.is_a?(Hash)
+          @log.warn 'Dropping log entries with malformed record: ' \
+                    "#{record.inspect}'. " \
```
I think you've lost the opening "`'`".

qingling128 replied:
Oops. Added.

qingling128 commented:
I cherry picked the other commits. Seems it still shows as `there is conflicts`. I'll rebase it after the code review is done.

PTAL.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -955,138 +980,175 @@ def determine_agent_level_common_labels
+
+      # Change the resource type and set matched_regexp_group if the tag matches
+      # certain regexp.
+      matched_regexp_group = nil # @tag_regexp_list can be an empty list.
```
Two spaces before the comment, please.

qingling128 replied:
Changed.

I tried `matched_regexp_group = nil<two_spaces_here># @tag_regexp_list` and Rubocop complains. So I guess you meant `matched_regexp_group = nil #<two_spaces_here> @tag_regexp_list`?

igorpeshansky replied:
I actually did mean two spaces before the `#`, but didn't realize Rubocop would complain. So let's go back to one space on each side of the `#`.

qingling128 replied:
Hahahaha, I see. Just changed it back.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1160,6 +1412,33 @@ def setup_prometheus
+    "#{DEFAULT_METADATA_AGENT_URL}/monitoredResource/#{local_resource_id}"
+  end
+
+  # Provide a stub context that initialize @logs_sent, execute the block and
```
Nit: initializes, executes, resets...

qingling128 replied:
Changed.

qingling128 commented:
Rebased. PTAL.

BTW, I fixed a few minor comments while rebasing. Also reverted my change to the Log Viewer URL since there is a separate bug to addresss that. Here are changes between the latest and the code before rebasing:

```
git diff origin/lingshi-container-rebase-backup..origin/lingshi-container

diff --git a/lib/fluent/plugin/out_google_cloud.rb b/lib/fluent/plugin/out_google_cloud.rb
index 3b81bc9..795c22c 100644
--- a/lib/fluent/plugin/out_google_cloud.rb
+++ b/lib/fluent/plugin/out_google_cloud.rb
@@ -393,12 +393,6 @@ module Fluent
       end

       # Set required variables: @project_id, @vm_id, @vm_name and @zone.
-      # If any info above is included in the response from Metadata Agent, make
-      # use of that. Otherwise make some additional requests to metadata server.
-      #
-      # Note: Once we support metadata injection on the Logging API side, we
-      # might no longer need to require all these metadata in logging agent. But
-      # for now, they are still required.
       set_required_metadata_variables

       # Retrieve monitored resource.
@@ -439,7 +433,8 @@ module Fluent

       # Log an informational message containing the Logs viewer URL
       @log.info 'Logs viewer address: https://console.cloud.google.com/logs/',
-                "viewer?project=#{@project_id}"
+                "viewer?project=#{@project_id}&resource=#{@resource.type}/",
+                "instance_id/#{@vm_id}"
     end

     def start

@@ -592,8 +585,7 @@ module Fluent
             end

           rescue GRPC::Cancelled => error
-            increment_failed_requests_count(
-              GRPC::Core::StatusCodes::CANCELLED)
+            increment_failed_requests_count(GRPC::Core::StatusCodes::CANCELLED)
             # RPC cancelled, so retry via re-raising the error.
             raise error

@@ -1125,7 +1117,7 @@ module Fluent
           @cloudfunctions_log_match =
             @compiled_cloudfunctions_log_regexp.match(record['log'])
           common_labels['execution_id'] =
-            @cloudfunctions_log_match['execution_id'] if \
+            @cloudfunctions_log_match['execution_id'] if
               @cloudfunctions_log_match &&
               @cloudfunctions_log_match['execution_id']
         end

diff --git a/test/plugin/base_test.rb b/test/plugin/base_test.rb
index 46fd6bd..5e570c7 100644
--- a/test/plugin/base_test.rb
+++ b/test/plugin/base_test.rb
@@ -1272,10 +1272,10 @@ module BaseTest
     end
   end

-  # Test logs from applications running in GKE. These logs have the label
+  # Test GKE container logs. These logs have the label
   # "logging.googleapis.com/local_resource_id" set in the format of
   # "gke_containerName.<namespace_id>.<pod_name>.<container_name>".
-  def test_gke_container_application_logs
+  def test_gke_container_logs
     [1, 2, 3, 5, 11, 50].each do |n|
       new_stub_context do
         setup_gce_metadata_stubs
@@ -1284,7 +1284,7 @@ module BaseTest
         setup_logging_stubs do
           d = create_driver(ENABLE_METADATA_AGENT_CONFIG)
           n.times do |i|
-            d.emit(gke_application_container_log_entry(log_entry(i)))
+            d.emit(gke_container_log_entry(log_entry(i)))
           end
           d.run
         end
@@ -1471,7 +1471,7 @@ module BaseTest
     }
   end

-  def gke_application_container_log_entry(log)
+  def gke_container_log_entry(log)
     {
       log: log,
       LOCAL_RESOURCE_ID_KEY =>
diff --git a/test/plugin/constants.rb b/test/plugin/constants.rb
index f354b65..8297486 100644
--- a/test/plugin/constants.rb
+++ b/test/plugin/constants.rb
@@ -601,7 +601,7 @@ module Constants
           'container_id' => DOCKER_CONTAINER_ID
         }
       }.to_json,
-    # GKE container application logs.
+    # GKE container logs.
     "gke_containerName.#{CONTAINER_NAMESPACE_ID}.#{CONTAINER_POD_NAME}." \
     "#{CONTAINER_CONTAINER_NAME}" =>
       {
```

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -420,39 +449,27 @@ def shutdown
         entries = []
-        group_resource, group_common_labels =
-          determine_group_level_monitored_resource_and_labels(tag)
+        group_level_resource, group_level_common_labels =
```
[Optional] I'm fine with keeping this as-is, but this renaming (`group_resource`->`group_level_resource`, `entry_resource`->`entry_level_resource`, etc) seems somewhat gratuitous to me.

qingling128 replied:
Ah, originally I renamed them because I added `tag_resource` and `tag_labels`. It's kind of ambiguous what that means. So i named them in an more explicit way showing that these are just regular monitored resource and metadata labels except they are at different levels.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -954,138 +974,175 @@ def determine_agent_level_common_labels
-      group_common_labels.freeze
-
-      [group_resource, group_common_labels]
+    # Group the log entries by tag and local_resource_id pairs.
```
`# Also filter out invalid non-Hash entries.`

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -954,138 +974,175 @@ def determine_agent_level_common_labels
+        unless retrieved_resource.nil?
+          # TODO(qingling128): Fix this temporary renaming from 'gke_container'
+          # to 'container'.
+          retrieved_resource.type = 'container' if
```
[Optional] If you move this line below the `resource` assignment, you could use `resource` instead of `retrieved_resource` and fit this on one line.

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -954,138 +974,175 @@ def determine_agent_level_common_labels
+          local_resource_id)
+        @log.debug 'Retrieved monitored resource from metadata agent: ' \
+                  "#{retrieved_resource.inspect}."
+        unless retrieved_resource.nil?
```
`if retrieved_resource`

qingling128 replied:
Changed.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -354,28 +381,30 @@ def configure(conf)
+        # The local_resource_id for this should be the instance id. Since this
+        # can be implicitly inferred by Metadata Agent, we do not need to
+        # explicitly send the key.
+        @resource = call_metadata_agent_for_monitored_resource(
```
Do we want add a TODO here to retrieve the resource on demand (similar to the one you have in line 1094)?

qingling128 replied:
Added.

igorpeshansky commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -1111,14 +1168,56 @@ def determine_entry_level_labels(group_resource, record)
+
+    # Call Metadata Agent to get monitored resource information and parse
+    # response to Google::Api::MonitoredResource.
+    def call_metadata_agent_for_monitored_resource(local_resource_id)
```
Same here: s/`call`/`query`/?

qingling128 replied:
Changed.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1106,6 +1106,197 @@ def test_log_entry_trace_field
+        d.emit(docker_container_application_log_entry(log_entry(0)))
+        d.run
+      end
+      verify_log_entries(1, DOCKER_CONTAINER_PARAMS_WITH_NO_STREAM) do |entry|
```
Careful: since you've merged #146, this now simply verifies that there's 1 entry with a 'textPayload' field, but no longer verifies the content of that field. See, e.g., `test_timestamps`, `test_one_container_log_from_tag_stderr`, and `verify_container_logs`.

qingling128 replied:
Ah, i see. Just added the corresponding `verify_default_log_entry_text` checks.

igorpeshansky replied:
Oops, sorry, I should have been clearer. This check happens automatically when you don't pass a block, so you only needed to change it here.

qingling128 replied:
Ah, I see. I should have taken a look at the change in #146. Just changed it back.

igorpeshansky commented on test/plugin/base_test.rb:
```diff
@@ -1106,6 +1106,197 @@ def test_log_entry_trace_field
+      end
+      verify_log_entries(1, DOCKER_CONTAINER_PARAMS_WITH_NO_STREAM) do |entry|
+        # Timestamp in 'time' field from log entry should be set properly.
+        assert_equal DOCKER_CONTAINER_SECONDS_EPOCH,
```
Other similar tests use `assert_equal_with_default` -- does it make sense to be consistent?

qingling128 replied:
Neither `DOCKER_CONTAINER_SECONDS_EPOCH` nor `DOCKER_CONTAINER_NANOS` is `0`, so we should be fine asserting it this way.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -83,26 +86,43 @@ module Constants
-        .map { |consts| [consts[:resource_type], consts[:metadata_attributes]] }
-        .to_h
+        [APPENGINE_CONSTANTS, GKE_CONSTANTS, DATAPROC_CONSTANTS]
+        .map do |consts|
```
Changed.

qingling128 commented on lib/fluent/plugin/out_google_cloud.rb:
```diff
@@ -467,16 +484,17 @@ def write(chunk)
-              record['time'] = timestamp
-            end
+            # Restore timestamp and severity if necessary. Note that the nested
+            # json might also has 'time' and 'severity' fields. If that is the
```
Changed.

qingling128 commented:
PTAL.

qingling128 commented:
Do we want to squash and merge everything in one commit? Or to keep refactoring, docker support and gke support in separate commits?

igorpeshansky commented:
Let's just squash and merge. The history will remain in this PR.
